src/main/java/week5/AlphabetFinder.java

```diff
@@ -0,0 +1,166 @@
+package week5;
+
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+
+public class AlphabetFinder {
+
+  /*
+   * The graph is represented using adjacency List as 1) Alphabet might be infinitely large, 2)
+   * from example the graph seem to be sparse 3) we do not check the presence of arbitrary edges
+   * but systematically parse through the graph.
+   *
+   * Main Steps: build graph from list of words -> topological sort
+   *
+   * Returns : empty <- if there is a cycle in the words
+   *           List<Character> <- the ordered alphabet
+   *
+   * Time Complexity: O(build) + O(sort)
+   *
+   *
+   */
+  public static Optional<List<Character>> findAlphabet(final String[] words) {
+
+    if (words == null) {
+      return Optional.empty();
+    }
+
+    Graph graph = buildGraph(words);
+    return topologicalSort(graph);
```

Comment:
would it make more sense to move topological sort to the `Graph` class as it is more a Graph operation and then this would be `buildGraph(words).topologicalSort();`